### PR TITLE
Invoke NullableWalker for all methods in DEBUG

### PR DESCRIPTION
**Pull request by cston**
Some existing `NullableWalker` tests are now skipped. Those tests will be re-enabled in a subsequent PR.

Extracted from #23981.

Please ignore the first commit which is PR #24901.

### CODE REVIEW

**Comment by cston** [comment]
@AlekseyTs, @jcouv, @dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -1932,8 +1821,25 @@ private static TypeSymbolWithAnnotations InferResultNullability(Conversion conve
                     var methodOpt = conversion.Method;
                     if ((object)methodOpt != null && methodOpt.ParameterCount == 1)
                     {
+                        if (Binder.RequiresLiftedUserDefinedConversion(methodOpt, operandType.TypeSymbol))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >Binder.RequiresLiftedUserDefinedConversion(methodOpt, operandType.TypeSymbol) [](start = 28, length = 77)

Is lifted flag not sufficient on the conversion node?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -1932,8 +1821,25 @@ private static TypeSymbolWithAnnotations InferResultNullability(Conversion conve
...
+                            break;
+                        }
                         // PROTOTYPE(NullableReferenceTypes): Update method based on operandType.
-                        return methodOpt.ReturnType;
+                        var resultType = methodOpt.ReturnType;
+                        // Ensure converted type matches expected ToType, specifically for lifted
+                        // conversions. See LocalRewriter.MakeConversionNode for similar handling.
+                        if (!fromConversionNode)
+                        {
+                            var toType = conversion.BestUserDefinedConversionAnalysis.ToType;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >toType [](start = 32, length = 6)

How is toType different from targetType?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -2093,23 +2002,27 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
...
             // byref assignment is also a potential write
             if (node.RefKind != RefKind.None)
             {
                 WriteArgument(node.Right, node.RefKind, method: null);
             }
 
-            TrackNullableStateForAssignment(node.Left, left.Slot, left.Type, node.Right, right.Type, right.Slot);
-
-            //if (this.State.Reachable) // PROTOTYPE(NullableReferenceTypes): Consider reachability?
+            if (node.Left.Kind == BoundKind.EventAccess && ((BoundEventAccess)node.Left).EventSymbol.IsWindowsRuntimeEvent)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >IsWindowsRuntimeEvent [](start = 101, length = 21)

What about regular not field-like events?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2). I assume the first commit is from another PR that I already reviewed and signed off on. Correct?

**Comment by cston** [comment]
@AlekseyTs, yes the first commit is the from the previous PR you reviewed and approved.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: It looks like `toType` can differ from `targetType` by tuple element names at least. (`CodeGenTupleTests.NullCoalesingOperatorWithTupleNames` contains an example where `methodOpt.ReturnType is `(int d1, int d2)`, `toType` is (int a, int c)?` and `targetType` is (int d1, int d2)?`). Should `InferResultNullability` simply return `TypeSymbolWithAnnotations(targetType)` if `methodOpt.ReturnType != targetType`?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Changed to `if (conversion.BestUserDefinedConversionAnalysis.Kind == UserDefinedConversionAnalysisKind.ApplicableInLiftedForm)`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Assignment of non-field-like events uses `BoundEventAssignmentOperator`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -1932,8 +1821,25 @@ private static TypeSymbolWithAnnotations InferResultNullability(Conversion conve
...
-                        return methodOpt.ReturnType;
+                        var resultType = methodOpt.ReturnType;
+                        // Ensure converted type matches expected target type, specifically for lifted
+                        // conversions. See LocalRewriter.MakeConversionNode for similar handling
+                        // based on conversion.BestUserDefinedConversionAnalysis.ToType.
+                        if (!fromConversionNode)
+                        {
+                            if (resultType.TypeSymbol != targetType)
+                            {
+                                // PROTOTYPE(NullableReferenceTypes): Should handle nested nullability.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >// PROTOTYPE(NullableReferenceTypes): Should handle nested nullability. [](start = 32, length = 71)

I think something should be done about top level nullability here as well, I suggest adjusting the comment. I think, for user-defined conversions, we should treat transition from resultType to target type as just another built-in conversion.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: > Assignment of non-field-like events uses BoundEventAssignmentOperator.

Consider adding a comment.

---
In reply to: [169229495](https://github.com/dotnet/roslyn/pull/24917#discussion_r169229495) [](ancestors = 169229495)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 5]
```diff
@@ -460,22 +468,29 @@ private void InheritNullableStateOfFieldOrProperty(int targetContainerSlot, int
...
-                        InheritNullableStateOfTrackableType(targetMemberSlot, valueSlot, isByRefTarget);
+                        InheritNullableStateOfTrackableType(targetMemberSlot, valueMemberSlot, isByRefTarget, slotWatermark);
                     }
                 }
             }
             else if (EmptyStructTypeCache.IsTrackableStructType(fieldOrPropertyType.TypeSymbol))
             {
-                var slot = GetOrCreateSlot(fieldOrProperty, targetContainerSlot);
-                if (slot > 0)
+                var targetMemberSlot = GetOrCreateSlot(fieldOrProperty, targetContainerSlot);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Nit: `int` type

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 4 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Added comment.

---
In reply to: [169409574](https://github.com/dotnet/roslyn/pull/24917#discussion_r169409574) [](ancestors = 169409574)

**Pull request merged**